src/core/qgsmaplayerrenderer.h

```diff
@@ -66,6 +67,13 @@ class CORE_EXPORT QgsMapLayerRenderer
     //! Gets access to the ID of the layer rendered by this class
     QString layerId() const { return mLayerID; }
 
+    /**
+     * Resets the painter to use for the job to a new \a painter.
```

Comment:
how it is with ownership of the painter?

question: don't we kind of assume the painter cannot be null in the renderers?

thinking loud: I am wondering if the rendering context should not be part of the `QgsMapLayerRenderer`, since now we have a virtual setter in base class that does not in principle contains at all the context/painter.

Reply:
>  I am wondering if the rendering context should not be part of the QgsMapLayerRenderer

it probably should be, yes...

Reply:
@PeterPetrik does c1d156d look better to you? I've moved the render context member to the QgsMapLayerRenderer base class. There's a further cleanup which should be done in QGIS 4, which is to make the base class member a reference rather than a pointer, but I can't do that without breaking stable API.

Reply:
looks right, thanks